File: lager1/stadiljus/submissions/wrong_answer/cheese_interval_greedy.py
#!/usr/bin/env python3
# @EXPECTED_GRADES@ WA WA WA WA WA
#
# Cheese: Computes farthest visible left/right for each position via the
# Cartesian tree, then treats coverage as a contiguous interval and does
# greedy interval covering. Visibility is NOT contiguous in general, so
# this overclaims coverage and underestimates the answer.

import sys


def main():
    input = sys.stdin.readline
    n = int(input())
    p = list(map(int, input().split()))

    if n == 1:
        print(1)
        return

    # Build Cartesian tree (max-heap)
    left = [-1] * n
    right = [-1] * n
    parent = [-1] * n
    stack = []

    for i in range(n):
        last = -1
        while stack and p[stack[-1]] < p[i]:
            last = stack.pop()
        if stack:
            right[stack[-1]] = i
            parent[i] = stack[-1]
        if last != -1:
            left[i] = last
            parent[last] = i
        stack.append(i)

    # For each node, compute farthest visible left and right
    # Visibility set of i: {parent[i], left[i], right[i], i-1, i+1} (where valid)
    # Interval = [min of visible positions, max of visible positions]
    intervals = []
    for i in range(n):
        lo = i
        hi = i
        if parent[i] != -1:
            lo = min(lo, parent[i])
            hi = max(hi, parent[i])
        if left[i] != -1:
            lo = min(lo, left[i])
        if right[i] != -1:
            hi = max(hi, right[i])
        if i > 0:
            lo = min(lo, i - 1)
        if i < n - 1:
            hi = max(hi, i + 1)
        intervals.append((lo, hi))

    # Greedy interval cover of [0, n-1]
    # Sort by left endpoint for the sweeping algorithm
    indexed = sorted(range(n), key=lambda i: intervals[i][0])

    answer = 0
    covered_until = -1
    idx = 0

    while covered_until < n - 1:
        best_right = -1
        while idx < n and intervals[indexed[idx]][0] <= covered_until + 1:
            best_right = max(best_right, intervals[indexed[idx]][1])
            idx += 1
        if best_right == -1:
            break
        covered_until = best_right
        answer += 1

    print(answer)


if __name__ == "__main__":
    main()

File: lager1/uppsikt/submissions/wrong_answer/cheese_interval_greedy.cpp
// @EXPECTED_GRADES@ WA WA WA WA WA
//
// Cheese: Computes farthest visible left/right for each position via the
// Cartesian tree, then treats coverage as a contiguous interval and does
// greedy interval covering. Visibility is NOT contiguous in general, so
// this overclaims coverage and underestimates the answer.

#include <bits/stdc++.h>
using namespace std;

int main() {
    int n;
    scanf("%d", &n);
    vector<int> p(n);
    for (int i = 0; i < n; i++) scanf("%d", &p[i]);

    if (n == 1) {
        printf("1\n");
        return 0;
    }

    // Build Cartesian tree (max-heap)
    vector<int> L(n, -1), R(n, -1), par(n, -1);
    vector<int> stk;

    for (int i = 0; i < n; i++) {
        int last = -1;
        while (!stk.empty() && p[stk.back()] < p[i]) {
            last = stk.back();
            stk.pop_back();
        }
        if (!stk.empty()) {
            R[stk.back()] = i;
            par[i] = stk.back();
        }
        if (last != -1) {
            L[i] = last;
            par[last] = i;
        }
        stk.push_back(i);
    }

    // For each node, compute interval [lo, hi]
    // Visibility: parent, left child, right child, i-1, i+1
    vector<pair<int,int>> intervals(n);
    for (int i = 0; i < n; i++) {
        int lo = i, hi = i;
        if (par[i] != -1) {
            lo = min(lo, par[i]);
            hi = max(hi, par[i]);
        }
        if (L[i] != -1) lo = min(lo, L[i]);
        if (R[i] != -1) hi = max(hi, R[i]);
        if (i > 0) lo = min(lo, i - 1);
        if (i < n - 1) hi = max(hi, i + 1);
        intervals[i] = {lo, hi};
    }

    // Sort by left endpoint
    vector<int> idx(n);
    iota(idx.begin(), idx.end(), 0);
    sort(idx.begin(), idx.end(), [&](int a, int b) {
        return intervals[a].first < intervals[b].first;
    });

    // Greedy interval cover of [0, n-1]
    int answer = 0;
    int covered_until = -1;
    int j = 0;

    while (covered_until < n - 1) {
        int best_right = -1;
        while (j < n && intervals[idx[j]].first <= covered_until + 1) {
            best_right = max(best_right, intervals[idx[j]].second);
            j++;
        }
        if (best_right == -1) break;
        covered_until = best_right;
        answer++;
    }

    printf("%d\n", answer);
    return 0;
}
